# Laboratorium 5
### Michał Kurleto

In [1]:
from mpmath import *
import matplotlib.pyplot as plt
import numpy as np

### Funkcje i ich pochodne: 

In [2]:
function1 = lambda x: cos(x) * cosh(x) - 1 
derivative1 = lambda x: sinh(x) * cos(x) - sin(x) * cosh(x)

function2 = lambda x: (1/x) - tan(x)
derivative2 = lambda x: -1/(x**2) - 1/(cos(x)**2)

function3 = lambda x: 2**(-x) + e**x + 2 * cos(x) - 6
derivative3 = lambda x: e**x - ln(2) * 2**(-x) - 2 * sin(x)

### 1. Metoda Bisekcji
Napisz funkcję realizującą metodę bisekcji dla danej funkcji f w oparciu o arytmetykę o
zmiennej precyzji (mpmath.mpf). Funkcja przyjmuje następujące argumenty:   
• Minimalną precyzję obliczeń (liczba cyfr znaczących)   
• Krańce przedziału   
• Błąd bezwzględny obliczeń1   

Funkcja ma zwracać wyznaczone miejsce zerowe, wyliczoną wartość w miejscu zerowym
oraz liczbę iteracji potrzebną do uzyskania określonej dokładności. Przetestuj działanie
metody dla funkcji podanych na początku instrukcji. Jaka liczba iteracji jest potrzebna do uzyskania bezwzględnej dokładności rzędu: 10−7
, 10−15, 10−33? 
W jaki sposób możemy obliczyć k pierwszych dodatnich pierwiastków funkcji f1(x)?


In [3]:
def bisection(function, left, right, epsilon, precision):
    mp.dps = precision
    iterations = 0
    
    while right - left >= epsilon:
        mid = left + (right - left)/2
        
        if function(left) * function(mid) < 0: #są różnych znaków
            right = mid
        elif function(right) * function(mid) < 0:
            left = mid
        else:
            break
        
        iterations += 1
    
    return mid, function(mid), iterations

#### Funkcja 1.

In [4]:
print(bisection(function1, (3/2) * pi,  2 * pi, mpf(1e-7), 7))
print(bisection(function1, (3/2) * pi,  2 * pi, mpf(1e-15), 15))
print(bisection(function1, (3/2) * pi,  2 * pi, mpf(1e-33), 33))

(mpf('4.730040789'), mpf('2.518296242e-6'), 24)
(mpf('4.7300407448627046'), mpf('3.5749181392930041e-14'), 51)
(mpf('4.73004074486270402602404810083388474'), mpf('-4.33334237487128066937595037081608647e-33'), 111)


#### Funkcja 2.

In [5]:
print(bisection(function2, mpf(1e-7),  pi/2, mpf(1e-7), 7))
print(bisection(function2, mpf(1e-15),  pi/2, mpf(1e-15), 15))
print(bisection(function2, mpf(1e-33),  pi/2, mpf(1e-33), 33))

(mpf('0.8603335917'), mpf('0.0'), 23)
(mpf('0.86033358901938017'), mpf('-1.3322676295501878e-15'), 51)
(mpf('0.860333589019379762483893424137662091'), mpf('9.62964972193617926527988971292463659e-34'), 111)


#### Funkcja 3.

In [6]:
print(bisection(function3, mpf(1),  mpf(3), mpf(1e-7), 7))
print(bisection(function3, mpf(1),  mpf(3), mpf(1e-15), 15))
print(bisection(function3, mpf(1),  mpf(3), mpf(1e-33), 33))

(mpf('1.829383552'), mpf('-1.788139343e-7'), 25)
(mpf('1.8293836019338494'), mpf('1.7763568394002505e-15'), 51)
(mpf('1.829383601933848817136212946814151'), mpf('0.0'), 110)


#### W jaki sposób możemy obliczyć k pierwszych dodatnich pierwiastków funkcji f1(x)?
Todo

### 2. Metoda Newtona
Napisz funkcję realizującą metodę Newtona w oparciu o arytmetykę o zmiennej precyzji
(mpmath.mpf). Funkcja ma wykorzystywać dwa kryteria stopu:   

• maksymalną liczbę iteracji   
• moduł różnicy kolejnych przybliżeń mniejszy od danej wartości ε   

Oprócz przybliżonej wartości pierwiastka funkcja ma zwrócić liczbę iteracji potrzebną
do uzyskania określonej dokładności ε. Przetestuj działanie metody dla funkcji podanych
na początku instrukcji (dodatkowo dostępne pochodne tych funkcji). Porównaj zbieżność
metody ze zbieżnością uzyskaną dla metody bisekcji.


In [7]:
def newton(function, derivative, left, right, epsilon, precision):
    mp.dps = precision
    
    iterations = 0
    x = left
    
    while abs((fx := function(x))) > epsilon:
        x -= fx / derivative(x)
        iterations += 1
        
    return x, fx, iterations 

#### Funkcja 1.

In [18]:
print(newton(function1, derivative1, (3/2) * pi, 2 * pi, mpf(1e-7), 8))
print(newton(function1, derivative1, (3/2) * pi, 2 * pi, mpf(1e-15), 17))
print(newton(function1, derivative1, (3/2) * pi, 2 * pi, mpf(1e-33), 35))

(mpf('4.7300407439'), mpf('-5.2154064178e-8'), 3)
(mpf('4.7300407448627040247'), mpf('-7.7195194680967915701e-17'), 4)
(mpf('4.7300407448627040260240481008338848171'), mpf('-1.6174802267314676109649814752178100527e-34'), 5)


#### Funkcja 2.

In [9]:
print(newton(function2, derivative2, mpf(0.1), pi/2, mpf(1e-7), 8))
print(newton(function2, derivative2, mpf(0.1), pi/2, mpf(1e-15), 17))
print(newton(function2, derivative2, mpf(0.1), pi/2, mpf(1e-33), 35))

(mpf('0.86033358984'), mpf('-3.7252902985e-9'), 6)
(mpf('0.86033358901937976269'), mpf('0.0'), 7)
(mpf('0.86033358901937976248389342413766233283'), mpf('1.5046327690525280101999827676444744676e-36'), 8)


#### Funkcja 3.

In [10]:
print(newton(function3, derivative3, mpf(1), mpf(3), mpf(1e-7), 8))
print(newton(function3, derivative3, mpf(1), mpf(3), mpf(1e-15), 17))
print(newton(function3, derivative3, mpf(1), mpf(3), mpf(1e-33), 35))

(mpf('1.8293836042'), mpf('7.4505805969e-9'), 7)
(mpf('1.8293836019338488158'), mpf('0.0'), 8)
(mpf('1.8293836019338488171362129468141507913'), mpf('0.0'), 9)


### 3. Metoda Siecznych
Napisz funkcję realizującą metodę siecznych w oparciu o arytmetykę o zmiennej precyzji
(mpmath.mpf). Funkcja powinna stosować te same kryteria stopu co funkcja realizująca
metodę Newtona. Przetestuj działanie metody dla funkcji podanych na początku instrukcji. Porównaj zbieżność metody ze zbieżnością uzyskaną dla metody bisekcji oraz
metody Newtona.

In [11]:
def secant(function, left, right, epsilon, precision):
    mp.dps = precision
    
    iterations = 0
    
    x0 = left
    x1 = right
    
    fx0 = function(x0)
    fx1 = function(x1)
    
    while abs(fx1) > epsilon:
        x = x1 - fx1 * (x1 - x0)/(fx1 - fx0)
        x1, x0 = x, x1
        
        fx0, fx1 = fx1, function(x)
        
        iterations += 1
    
    return x1, fx1, iterations

#### Funkcja 1.

In [12]:
print(secant(function1, (3/2) * pi, 2 * pi, mpf(1e-7), 8))
print(secant(function1, (3/2) * pi, 2 * pi, mpf(1e-15), 17))
print(secant(function1, (3/2) * pi, 2 * pi, mpf(1e-33), 35))

(mpf('4.7300407439'), mpf('-5.2154064178e-8'), 5)
(mpf('4.7300407448627040247'), mpf('-7.7195194680967915701e-17'), 7)
(mpf('4.7300407448627040260240481008338848231'), mpf('1.8506983059346094525459788042027035952e-34'), 8)


#### Funkcja 2.

In [14]:
print(secant(function2, mpf(0.1), pi/2, mpf(1e-7), 10))
print(secant(function2, mpf(0.1), pi/2, mpf(1e-15), 17))
print(secant(function2, mpf(0.1), pi/2, mpf(1e-33), 35))

(mpf('0.8603335884764'), mpf('2.008164301515e-9'), 10)
(mpf('0.86033358901937976269'), mpf('0.0'), 12)
(mpf('0.86033358901937976248389342413766233358'), mpf('-1.5046327690525280101999827676444744676e-36'), 13)


#### Funkcja 3.

In [16]:
print(secant(function3, mpf(1), mpf(3), mpf(1e-7), 8))
print(secant(function3, mpf(1), mpf(3), mpf(1e-15), 17))
print(secant(function3, mpf(1), mpf(3), mpf(1e-33), 35))

(mpf('1.8293836024'), mpf('0.0'), 9)
(mpf('1.8293836019338488141'), mpf('-6.9388939039072283776e-18'), 10)
(mpf('1.8293836019338488171362129468141507883'), mpf('-1.2037062152420224081599862141155795741e-35'), 12)


### 4. Porównanie Metod

#### Dla 7 miejsc znaczących:

|  Metoda |  Wynik |  Iteracje |
|---|---|---|
| Bisekcja:  | 4.7300407  | 24   |
|  Newton: |  4.7300407 |  3 |
|  Sieczne: | 4.7300407  |  5 |

#### Dla 15 miejsc znaczących:

|  Metoda |  Wynik |  Iteracje |
|---|---|---|
| Bisekcja:  | 4.73004074486270  | 51   |
|  Newton: |  4.73004074486270 |  4 |
|  Sieczne: | 4.73004074486270  |  7 |

#### Dla 33 miejsc znaczących:

|  Metoda |  Wynik |  Iteracje |
|---|---|---|
| Bisekcja:  | 4.73004074486270402602404810083388  | 111   |
|  Newton: |  4.73004074486270402602404810083388 |  5 |
|  Sieczne: | 4.73004074486270402602404810083388  |  8 |

### 5. Wnioski:
- **Bisekcja** - wymaga największej ilośći iteracji, ale za to gwarantuje że miejsce zerowe zostanie odnalezione.
- **Metoda Newtona** - ma większe wymagania co do funkcji, której miejsc zerowych szukamy. Możemy trafić na punkt, gdzie pochodna funkcji się zeruje co kończy się niepoprawnym wynikiem (błąd). Potrzebuje mniej iteracji niż metoda bisekcji.
- **Metoda Siecznych** - podobnie jak w metodzie Newtona, możne zwrócić niepoprawny wynik. Liczba iteracji zbliżona do tej w metodzie Newtona.